tsuru/client/plugin.go

```diff
@@ -57,37 +62,208 @@ func (c *PluginInstall) Run(context *cmd.Context, client *cmd.Client) error {
 	pluginName := context.Args[0]
 	pluginURL := context.Args[1]
 	if err := installPlugin(pluginName, pluginURL); err != nil {
-		return err
+		return fmt.Errorf("Error installing plugin %q: %w", pluginName, err)
 	}
 
 	fmt.Fprintf(context.Stdout, `Plugin "%s" successfully installed!`+"\n", pluginName)
 	return nil
 }
 
 func installPlugin(pluginName, pluginURL string) error {
-	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
-	file, err := filesystem().OpenFile(pluginPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+	tmpDir, err := filesystem().MkdirTemp("", "")
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not create a tmpdir: %w", err)
 	}
+	defer filesystem().RemoveAll(tmpDir)
+
 	resp, err := http.Get(pluginURL)
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not GET %q: %w", pluginURL, err)
 	}
+
 	defer resp.Body.Close()
-	data, err := ioutil.ReadAll(resp.Body)
+
+	data, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return err
 	}
 	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
 		return fmt.Errorf("Invalid status code reading plugin: %d - %q", resp.StatusCode, string(data))
 	}
-	n, err := file.Write(data)
+
+	// Try to extract .tar.gz first, then .zip. Fallbacks to copy the content
+	extractErr := extractTarGz(tmpDir, bytes.NewReader(data))
+	if extractErr != nil {
+		extractErr = extractZip(tmpDir, bytes.NewReader(data))
+	}
+	if extractErr != nil {
+		file, err := filesystem().OpenFile(filepath.Join(tmpDir, pluginName), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+		if err != nil {
+			return fmt.Errorf("Failed to open file: %w", err)
+		}
+		defer file.Close()
+		n, err := file.Write(data)
+		if err != nil {
+			return fmt.Errorf("Failed to write file content: %w", err)
+		}
+		if n != len(data) {
+			return fmt.Errorf("Incomplete write")
+		}
+	}
+
+	executablePath := findExecutablePlugin(tmpDir, pluginName)
+	if executablePath == "" {
+		return fmt.Errorf("The downloaded plugin content is invalid.")
+	}
+
+	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
+	if extractErr == nil {
+		os.Chmod(tmpDir, 0755)
+		if _, err := filesystem().Stat(pluginPath); err == nil {
+			filesystem().RemoveAll(pluginPath)
+		}
+		if err := filesystem().Rename(tmpDir, pluginPath); err != nil {
```

Comment:
```suggestion
		if err := filesystem().Rename(executablePath, pluginPath); err != nil {
```

I think we need just moving the executable file.

Reply:
My thoughts on this, is being able to install a plugin as an entire directory.
A tsuru plugin could then be made of multiple files, instead of a single file.

On the rpaasv2 example, the binary is accompanied by a README, which I find nice.
But this opens door to installing a python plugin, with multiple files for better organization, giving freedom to plugin's devs.

(this is why the `findExecutablePlugin()` was created, including sub directories of `~/.tsuru/plugins`)

Do you agree?

Reply:
Hum.. got it. I thought only to install the plugin executable itself.